File: 0367/solution.java
class Solution {
    public boolean isPerfectSquare(int num) {
        long left = 1, right = num;
        while (left <= right) {
            long pivot = left + (right - left) / 2;
            if (pivot * pivot == num)
                return true;
            if (pivot * pivot < num) {
                left = pivot + 1;
            } else {
                right = pivot - 1;
            }
        }
        return false;
    }
}

File: 0367/solution.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        left, right = 1, num
        while left <= right:
            pivot = (left + right) // 2
            square = pivot * pivot
            if square == num:
                return True
            if square > num:
                right = pivot - 1
            else:
                left = pivot + 1
        return False

        # time O(log(n))
        # space O(1)
